fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -952,33 +972,48 @@ def _wait_for_post_backward(self) -> None:
                 m.training_state = TrainingState.IDLE
 
     @torch.no_grad()
-    def _rebuild_full_params(self) -> None:
-        """Gather all shards of params."""
+    def _rebuild_full_params(self, full_precision: bool = False) -> None:
+        """
+        Gather all shards of params.
+
+        Args:
+            full_precision (bool, Optional): by default params will be gathered
+                in ``compute_dtype`` (e.g., FP16), unless *full_precision* is
+                ``True``, in which case they will be gathered in full precision
+                (e.g., FP32)
+        """
         with torch.cuda.stream(self._streams["all_gather"]):
-            if self.mixed_precision:
+            if self.mixed_precision and not full_precision:
                 self._cast_fp32_param_shards_to_fp16()
 
             for p in self.params:
-                if not p._is_sharded:
-                    if self.mixed_precision:
+                if not p._is_sharded:  # e.g., when world_size == 1
+                    if self.mixed_precision and not full_precision:
                         p.data = p._fp16_shard
                     continue
 
-                p_size = p._full_param_padded.size()
-                if p._full_param_padded.storage().size() != p_size.numel():
-                    # Allocate based on full size from all shards.
-                    alloc_storage_(p._full_param_padded, size=p_size)
-                    assert p_size.numel() % self.world_size == 0
-                    if p._is_sharded:
-                        # Fill p._full_param_padded with (p.data for each shard in self.world_size)
-                        chunks = list(p._full_param_padded.chunk(self.world_size))
-                        dist.all_gather(chunks, p.data, group=self.process_group)
-                    else:
-                        p._full_param_padded.copy_(torch.flatten(p.data), non_blocking=True)
-
-                p.data = p._full_param_padded[: p._orig_size.numel()].view(p._orig_size)
+                # If self.cpu_offload and full_precision, we need to cast the
+                # FP32 CPU param to CUDA for the all-gather.
+                p_data = p.data.to(p._full_param_padded.device)
 
-                if self.mixed_precision:
+                p_size = p._full_param_padded.size()
+                assert p_size.numel() % self.world_size == 0
+                if not self.mixed_precision or not full_precision:
+                    if p._full_param_padded.storage().size() != p_size.numel():
+                        # Allocate based on full size from all shards.
+                        alloc_storage_(p._full_param_padded, size=p_size)
+                    output_tensor = p._full_param_padded
+                else:
+                    # Allocate fresh tensor in full precision.
+                    output_tensor = p_data.new_zeros(p_size)
```

Comment:
these tensors will linger in GPU memory after use? Should we add a function to free them or make _free_full_params smarter in dealing with them?

Reply:
Good question, I should test this. My hope was that they there would only be one reference to them (in the returned `state_dict`), so it would be up to the caller to decide how/when to free them.

In particular, the `_use_fp32_param_shard` at the end of `summon_full_params` should remove the `p.data` reference added here, so the only remaining reference should be `state_dict`. But definitely worth testing :)